libbroker/broker/internal/core_actor.cc

```diff
@@ -355,77 +375,89 @@ caf::behavior core_actor_state::make_behavior() {
-          return f(xs, msg);
-        })
-        .compose(local_subscriber_scope_adder())
-        .subscribe(std::move(snk));
-    },
-    [this](std::shared_ptr<filter_type> fptr, data_producer_res snk) {
-      // Here, we accept a shared_ptr to the filter instead of an actual object.
-      // This allows the subscriber to manipulate the filter later by sending us
-      // an update message. The filter itself is not thread-safe. Hence, the
-      // publishers should never write to it directly.
-      subscribe(*fptr);
-      data_outputs
-        .filter([fptr = std::move(fptr)](const data_message& msg) {
-          detail::prefix_matcher f;
-          return f(*fptr, msg);
-        })
-        .compose(local_subscriber_scope_adder())
-        .subscribe(std::move(snk));
-    },
-    [this](std::shared_ptr<filter_type>& fptr, topic& x, bool add,
-           std::shared_ptr<std::promise<void>>& sync) {
-      // We assume that fptr belongs to a previously constructed flow.
-      auto e = fptr->end();
-      auto i = std::find(fptr->begin(), e, x);
-      if (add) {
-        if (i == e) {
-          fptr->emplace_back(std::move(x));
-          subscribe(*fptr);
-        }
-      } else {
-        if (i != e)
-          fptr->erase(i);
+    // -- interface for hubs ---------------------------------------------------
+    [this](hub_id id, filter_type& filter) {
+      // Update the filter of an existing hub.
+      auto i = hubs.find(id);
+      if (i == hubs.end()) {
+        log::core::error("update-hub-filter",
+                         "cannot update filter of hub {}: not found",
+                         static_cast<uint64_t>(id));
+        return;
+      }
+      auto& ptr = i->second;
+      if (ptr->filter != filter) {
+        log::core::debug("update-hub-filter", "update filter of hub {} to {}",
+                         static_cast<uint64_t>(id), filter);
+        ptr->filter = std::move(filter);
+        subscribe(ptr->filter);
       }
-      if (sync)
-        sync->set_value();
     },
-    // -- interface for publishers ---------------------------------------------
-    [this](data_consumer_res src) {
-      auto consumer_id = endpoint_id::random();
-      auto [in, sub] =
-        self
-          ->make_observable() //
-          .from_resource(std::move(src))
-          .do_on_next([this](const data_message&) {
-            metrics_for(packed_message_type::data).buffered->Increment();
-          })
-          .map([this](const data_message& msg) { return node_message{msg}; })
-          .compose(local_publisher_scope_adder())
-          .compose(add_killswitch_t{});
-      flow_inputs.push(in.do_finally([this, consumer_id] {
-                           auto i = subscriptions.find(consumer_id);
-                           if (i != subscriptions.end()) {
-                             subscriptions.erase(i);
-                           }
+    [this](hub_id id, filter_type& filter, bool filter_local,
```

Comment:
I think it'd be good to comment what `filter_local` means or use a name like `tap_local_hub_messages` or some such...

Reply:
I've added a comment.